### PR TITLE
Support for Discord RPC

### LINKED ISSUE #1109: Implement Discord Rich Presence into osu!
Rich Presence allows you to leverage the totally overhauled "Now Playing" section in a Discord user's profile to help people play your game together. Rich game data—including duration, score, current boss or map, and so much more—lives inside a user's profile alongside buttons that let you spectate or join a friend's game directly from Discord, or get a notification when it's over.

Rich Presence Image

_NOTE: The "Notify Me" feature is coming soon._

All you need is Discord's header file, and Discord's library.
In the header file, you'll find five event-emitting callbacks:

- ready()
- errored()
- disconnected()
- joinGame()
- spectateGame()

These five callbacks make up the entirety of what you need to implement. Behind the scenes, they do all the heavy lifting for you.

Learn more [here](https://discordapp.com/developers/docs/topics/rich-presence).

### PR DESCRIPTION
**Pull request by naoei**
Closes #1109 

Notes: you will have to change the ClientID to the official osu! discord client id, and use these corresponding names: https://i.imgur.com/a800eFi.png.
This could be optimized but im not completely sure.

Preview: https://u.teknik.io/01yQ3.mp4

### CODE REVIEW

File: osu.Game/Screens/Menu/MainMenu.cs [hunk 1]
```diff
@@ -44,6 +46,16 @@ public class MainMenu : OsuScreen
 
         protected override BackgroundScreen CreateBackground() => background;
 
+        protected override RichPresence Presence => new RichPresence
+        {
+            Details = "Idling",
+            Assets = new Assets()
```
Comment (peppy) [inline osu.Game/Screens/Menu/MainMenu.cs]: This cannot be repeated each usage. Making a base implementation in OsuScreen and only changing what is required is one fix but likely not what we want either.

File: osu.Game/Screens/Menu/MainMenu.cs [hunk 2]
```diff
@@ -188,6 +200,7 @@ public override void OnResuming(IScreen last)
             base.OnResuming(last);
 
             (Background as BackgroundScreenDefault)?.Next();
+            DiscordRpc.updatePresence(Presence);
```
Comment (peppy) [inline osu.Game/Screens/Menu/MainMenu.cs]: This should not be here.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
You added back a submodule.

Also this should likely be implemented in its own non-static component at `OsuGame` level, fetched via DI when the state is to be changed by a specific use case.

File: osu.Game/Utils/DiscordRPC.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+using DiscordRPC;
+using DiscordRPC.Logging;
+using LogLevel = DiscordRPC.Logging.LogLevel;
+
+namespace osu.Game.Utils
+{
+    public static class DiscordRpc
+    {
+        public static DiscordRpcClient client = new DiscordRpcClient("559391129716391967");
```
Comment (smoogipoo) [inline osu.Game/Utils/DiscordRPC.cs]:
```suggestion
        private static DiscordRpcClient client = new DiscordRpcClient("559391129716391967");
```

File: osu.Game/Utils/DiscordRPC.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+        /// <summary>Initializes and changes the Rich Presence in Discord.</summary>
+        /// <param name="Presence">Value to set the presence</param>
+        public static void updatePresence(RichPresence Presence)
```
Comment (smoogipoo) [inline osu.Game/Utils/DiscordRPC.cs]: Should be `UpdatePresence`.

File: osu.Game/Screens/Menu/MainMenu.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game/Screens/Menu/MainMenu.cs]: Since it's repeated pretty much every time in `OnResuming()`, it should be moved to `OsuScreen.OnResuming()`.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Just wanted to make sure you're aware that not all the review has been applied yet.

**Comment by AltenGD** [comment]
> Just wanted to make sure you're aware that not all the review has been applied yet.

oh sorry, i didn't notice the others

**Comment by AltenGD** [comment]
Ok, should be ready to go, if i missed anything please tell me

**Comment by peppy** [comment]
The same rich presence can't be specified in each screen. In fact it probably shouldn't be a RichPresence object but just enough metadata that the underlying discord rich presence can source from.

Basically we can't have discord specific code all over the code base. It should be something custom which is consumable by discord and/or other services (an interface with `Description` or `Status` strings).

**Comment by AltenGD** [comment]
Closed as #4654 is a better solution

**Pull request closed without merging**